### PR TITLE
Implement test for graceful restart of the primary unit

### PR DESCRIPTION
**Pull request by shayancanonical**
# Issue
We have not yet implemented the test for a graceful restart of the primary unit.

# Solution
Implement the test

# Release Notes
Implement test for graceful restart of the primary unit

### CODE REVIEW

File: tests/integration/high_availability/test_self_healing.py [hunk 1]
```diff
@@ -194,3 +194,62 @@ async def test_freeze_db_process(ops_test: OpsTest, continuous_writes) -> None:
     ), "The deployed mysql application does not have three online nodes"
 
     await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.self_healing_tests
+async def test_graceful_crash_of_priamry(ops_test: OpsTest, continuous_writes) -> None:
```
Comment (paulomach) [inline tests/integration/high_availability/test_self_healing.py:202]: nit: 
```suggestion
async def test_graceful_crash_of_primary(ops_test: OpsTest, continuous_writes) -> None:
```

File: tests/integration/high_availability/test_self_healing.py [hunk 2]
```diff
@@ -194,3 +194,62 @@ async def test_freeze_db_process(ops_test: OpsTest, continuous_writes) -> None:
...
+        for unit in ops_test.model.applications[mysql_application_name].units
+        if unit.name != primary.name
+    ]
+
+    # retring as it may take time for the cluster to recognize that the primary process is stopped
+    for attempt in Retrying(stop=stop_after_delay(2 * 60), wait=wait_fixed(10)):
+        with attempt:
+            assert await ensure_n_online_mysql_members(
+                ops_test, 3
+            ), "The deployed mysql application does not have two online nodes"
```
Comment (paulomach) [inline tests/integration/high_availability/test_self_healing.py:245]:
```suggestion
            ), "The deployed mysql application does not have three online nodes"
```

**Review by paulomach — APPROVED** [review]
lgtm

File: tests/integration/high_availability/test_self_healing.py (hunk 1 above)
Comment (shayancanonical) [reply tests/integration/high_availability/test_self_healing.py]: addressed in [fc75968](https://github.com/canonical/mysql-k8s-operator/pull/112/commits/fc759680f8d5ac7b94e46fdcfa6fa8fd4810863b)

File: tests/integration/high_availability/test_self_healing.py (hunk 2 above)
Comment (shayancanonical) [reply tests/integration/high_availability/test_self_healing.py]: addressed in [fc75968](https://github.com/canonical/mysql-k8s-operator/pull/112/commits/fc759680f8d5ac7b94e46fdcfa6fa8fd4810863b)

**Review by marceloneppel — APPROVED** [review]
Looks good!

**Review by MiaAltieri — APPROVED** [review]
Awesome work Shayan :)

**Pull request merged**